### PR TITLE
[TikTok Audiences] Add create audience

### PR DESCRIPTION
**Pull request by maryamsharif**
This PR adds in the create audience flow for TikTok Audiences. Users can either input an `audience_id` to sync to or not and we will create an audience based on the `audience_key`. Before we could not include the creation of audiences because centrifuge runs events in parallel and we would run the risk of a race condition causing 2 duplicate audiences being created.

This `integrations-go` [PR](https://github.com/segmentio/integrations-go/pull/364) sets concurrency for TikTok to 1 so we do not have to worry about the race condition.

## Testing

_Include any additional information about the testing you have completed to
ensure your changes behave as expected. For a speedy review, please check
any of the tasks you completed below during your testing._

- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [x] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [x] [Segmenters] Tested in the staging environment

### CODE REVIEW

File: packages/destination-actions/src/destinations/tiktok-audiences/addUser/index.ts [hunk 1]
```diff
@@ -46,9 +50,11 @@ const action: ActionDefinition<Settings, Payload> = {
     }
   },
   perform: async (request, { settings, payload }) => {
+    console.log('Processing payload...')
```
Comment (rhall-twilio) [inline packages/destination-actions/src/destinations/tiktok-audiences/addUser/index.ts:53]: Did you mean to keep these `console.log` ?

File: packages/destination-actions/src/destinations/tiktok-audiences/functions.ts [hunk 2]
```diff
@@ -36,14 +44,61 @@ export async function processPayload(
     // Segment just created isn't available yet for updates via this endpoint.
     // Audiences are usually available to accept batches of data 1 - 2 minutes after
     // they're created. Here, we'll throw an error and let Centrifuge handle the retry.
-    if (res.status !== 200) {
+    if (res.status === 400 && !payloads[0].audience_id && res.data.message.includes('please send valid audience ID')) {
       throw new RetryableError('Error while attempting to update TikTok Audience. This batch will be retried.')
```
Comment (rhall-twilio) [inline packages/destination-actions/src/destinations/tiktok-audiences/functions.ts:48]: Now that the error condition is specialized, the thrown error message can be less generic.

File: packages/destination-actions/src/destinations/tiktok-audiences/functions.ts [hunk 3]
```diff
@@ -17,10 +18,17 @@ export async function processPayload(
...
+    const audiences = await getAllAudiences(TikTokApiClient)
+    audience_id = await getAudienceID(TikTokApiClient, payloads[0], audiences)
+  } else {
+    audience_id = payloads[0].audience_id
+  }
 
   const id_schema = getIDSchema(payloads[0])
 
-  const users = extractUsers(payloads)
+  const users = extractUsers(payloads, audience_id)
```
Comment (rhall-twilio) [inline packages/destination-actions/src/destinations/tiktok-audiences/functions.ts:31]: Is there a reason the validation of audience_id happens in line 47 before we use it here?

File: packages/destination-actions/src/destinations/tiktok-audiences/functions.ts [hunk 4]
```diff
@@ -36,14 +44,61 @@ export async function processPayload(
...
       throw new RetryableError('Error while attempting to update TikTok Audience. This batch will be retried.')
     }
   }
 
   return res
 }
 
+// TikTok returns a max of 100 audiences per request to their `list` endpoint.
+// A customer can have a max of 400 audiences in a single advertiser account.
+// We may have to make up to 4 requests to get all audiences.
```
Comment (rhall-twilio) [inline packages/destination-actions/src/destinations/tiktok-audiences/functions.ts:57]: Do you know if we can fulfill a request of this kind in the allotted time? If not, we may want to artificially limit the number of audiences we're willing to process.

File: packages/destination-actions/src/destinations/tiktok-audiences/addUser/index.ts [hunk 5]
```diff
@@ -46,9 +50,11 @@ const action: ActionDefinition<Settings, Payload> = {
...
     return processPayload(request, settings, [payload], 'add')
   },
   performBatch: async (request, { settings, payload }) => {
+    console.log('Processing batch payload...')
```
Comment (marinhero) [inline packages/destination-actions/src/destinations/tiktok-audiences/addUser/index.ts:57]: And this one too 😬

File: packages/destination-actions/src/destinations/tiktok-audiences/functions.ts [hunk 6]
```diff
@@ -17,10 +18,17 @@ export async function processPayload(
 
   const selected_advertiser_id = payloads[0].selected_advertiser_id ?? undefined
   const TikTokApiClient: TikTokAudiences = new TikTokAudiences(request, selected_advertiser_id)
+  let audience_id
+  if (!payloads[0].audience_id) {
```
Comment (marinhero) [inline packages/destination-actions/src/destinations/tiktok-audiences/functions.ts:22]: You are testing for `undefined` in the other variables (line 19) do the same for this one.

File: packages/destination-actions/src/destinations/tiktok-audiences/addUser/index.ts (hunk 1 above)
Comment (maryamsharif) [reply packages/destination-actions/src/destinations/tiktok-audiences/addUser/index.ts:53]: Whoops, meant to take it out, thank you!

File: packages/destination-actions/src/destinations/tiktok-audiences/functions.ts [hunk 7]
```diff
@@ -36,14 +44,61 @@ export async function processPayload(
...
   return res
 }
 
+// TikTok returns a max of 100 audiences per request to their `list` endpoint.
+// A customer can have a max of 400 audiences in a single advertiser account.
+// We may have to make up to 4 requests to get all audiences.
+// The first request will return the total_number of audiences associated with
+// the advertiser account.
+export async function getAllAudiences(TikTokApiClient: TikTokAudiences) {
+  let response = await TikTokApiClient.getAudiences(1, 100)
```
Comment (marinhero) [inline packages/destination-actions/src/destinations/tiktok-audiences/functions.ts:61]: What do 1,100 means?

File: packages/destination-actions/src/destinations/tiktok-audiences/functions.ts [hunk 8]
```diff
@@ -36,14 +44,61 @@ export async function processPayload(
...
+  let audiences: Audiences[] = response.data.list
+  const total_number_audiences = response.data.page_info.total_number
+  let recieved_audiences = response.data.page_info.page_size
+  let page_number = 2
```
Comment (marinhero) [inline packages/destination-actions/src/destinations/tiktok-audiences/functions.ts:65]: consider a meaningful name so the code conveys its behavior.
```suggestion
  let page_number = PAGINATION_START_INDEX
```

File: packages/destination-actions/src/destinations/tiktok-audiences/functions.ts [hunk 9]
```diff
@@ -36,14 +44,61 @@ export async function processPayload(
...
+      return audience.audience_id
+    }
+  })
+
+  // More than 1 audience returned matches name
+  if (audienceExists.length > 1) {
+    throw new IntegrationError('Multiple audiences found with the same name', 'INVALID_SETTINGS', 400)
+  }
+
+  if (audienceExists.length == 1) {
```
Comment (marinhero) [inline packages/destination-actions/src/destinations/tiktok-audiences/functions.ts:92]:
```suggestion
  if (audienceExists.length === 1) {
```

File: packages/destination-actions/src/destinations/tiktok-audiences/functions.ts [hunk 10]
```diff
@@ -36,14 +44,61 @@ export async function processPayload(
...
+    audienceID = audienceExists[0].audience_id
+  } else {
+    const response = await TikTokApiClient.createAudience(payload)
+    audienceID = response.data.data.audience_id
```
Comment (marinhero) [inline packages/destination-actions/src/destinations/tiktok-audiences/functions.ts:96]: `data.data` ? 🥲

**Pull request closed without merging**